File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/TankSessionLifecycle.kt
package com.rork.vinetrack.data

import com.rork.vinetrack.data.model.TankSession
import com.rork.vinetrack.data.model.Trip

/** Deterministic production transitions for operator-driven tank spraying. */
object TankSessionLifecycle {
    fun start(
        trip: Trip,
        timestamp: String,
        currentRow: Double?,
        makeId: () -> String = { java.util.UUID.randomUUID().toString() },
    ): Trip {
        if (trip.activeTankNumber != null) return trip

        val sessions = trip.tankSessions.toMutableList()
        val fillIndex = reusableFillOnlySessionIndex(trip)
        val activeTankNumber: Int
        if (fillIndex >= 0) {
            val existing = sessions[fillIndex]
            sessions[fillIndex] = existing.copy(
                startTime = timestamp,
                startRow = currentRow,
                endTime = null,
                endRow = null,
                fillEndTime = existing.fillEndTime ?: timestamp,
            )
            activeTankNumber = existing.tankNumber
        } else {
            activeTankNumber = (sessions.maxOfOrNull { it.tankNumber } ?: 0) + 1
            sessions.add(
                TankSession(
                    id = makeId(),
                    tankNumber = activeTankNumber,
                    startTime = timestamp,
                    startRow = currentRow,
                ),
            )
        }
        return trip.copy(
            tankSessions = sessions,
            activeTankNumber = activeTankNumber,
            isFillingTank = false,
            fillingTankNumber = null,
        )
    }

    fun end(trip: Trip, timestamp: String, currentRow: Double?): Trip {
        val activeTankNumber = trip.activeTankNumber ?: return trip
        val sessionIndex = trip.tankSessions.indexOfLast {
            it.tankNumber == activeTankNumber && it.endTime == null && it.startRow != null
        }
        if (sessionIndex < 0) return trip

        val sessions = trip.tankSessions.toMutableList()
        sessions[sessionIndex] = sessions[sessionIndex].copy(endTime = timestamp, endRow = currentRow)
        return trip.copy(tankSessions = sessions, activeTankNumber = null)
    }

    private fun reusableFillOnlySessionIndex(trip: Trip): Int {
        fun TankSession.isFillOnly(): Boolean = fillStartTime != null && endTime == null && startRow == null
        val fillingNumber = trip.fillingTankNumber
        if (fillingNumber != null) {
            val matching = trip.tankSessions.indexOfLast { it.tankNumber == fillingNumber && it.isFillOnly() }
            if (matching >= 0) return matching
        }
        return trip.tankSessions.indexOfLast { it.isFillOnly() }
    }
}

File: ios/VineTrack/LegacyImported/Services/TankSessionLifecycle.swift
import Foundation

/// Deterministic production transitions for operator-driven tank spraying.
nonisolated enum TankSessionLifecycle {
    static func start(
        trip: Trip,
        at timestamp: Date,
        currentRow: Double,
        makeID: () -> UUID = UUID.init
    ) -> Trip {
        guard trip.activeTankNumber == nil else { return trip }

        var updated = trip
        if let fillIndex = reusableFillOnlySessionIndex(in: updated) {
            let tankNumber = updated.tankSessions[fillIndex].tankNumber
            updated.tankSessions[fillIndex].startTime = timestamp
            updated.tankSessions[fillIndex].startRow = currentRow
            updated.tankSessions[fillIndex].endTime = nil
            updated.tankSessions[fillIndex].endRow = nil
            if updated.tankSessions[fillIndex].fillEndTime == nil {
                updated.tankSessions[fillIndex].fillEndTime = timestamp
            }
            updated.activeTankNumber = tankNumber
        } else {
            let nextNumber = (updated.tankSessions.map(\.tankNumber).max() ?? 0) + 1
            updated.tankSessions.append(TankSession(
                id: makeID(),
                tankNumber: nextNumber,
                startTime: timestamp,
                startRow: currentRow
            ))
            updated.activeTankNumber = nextNumber
        }
        updated.isFillingTank = false
        updated.fillingTankNumber = nil
        return updated
    }

    static func end(trip: Trip, at timestamp: Date, currentRow: Double) -> Trip {
        guard let activeTankNumber = trip.activeTankNumber else { return trip }
        guard let sessionIndex = trip.tankSessions.lastIndex(where: {
            $0.tankNumber == activeTankNumber && $0.endTime == nil && $0.startRow != nil
        }) else { return trip }

        var updated = trip
        updated.tankSessions[sessionIndex].endTime = timestamp
        updated.tankSessions[sessionIndex].endRow = currentRow
        updated.activeTankNumber = nil
        return updated
    }

    private static func reusableFillOnlySessionIndex(in trip: Trip) -> Int? {
        let isFillOnly: (TankSession) -> Bool = {
            $0.fillStartTime != nil && $0.endTime == nil && $0.startRow == nil
        }
        if let fillingTankNumber = trip.fillingTankNumber,
           let index = trip.tankSessions.lastIndex(where: {
               $0.tankNumber == fillingTankNumber && isFillOnly($0)
           }) {
            return index
        }
        return trip.tankSessions.lastIndex(where: isFillOnly)
    }
}
